File: Programmers/JiHo/5주차(DFS:BFS)/점프와 순간 이동.py
'''
(현재까지 온 거리) x 2 조건으로 이진수를 구하는 과정과 매우 흡사하다고 판단하였다.
맨 처음 풀이에서 스택을 이용하여 풀이하였으나 스택을 개선하지 않는 방법으로 개선하였다.
'''
def solution(n):
    stack = []
    while True : #스택에 이진수 넣기
        stack.append(n % 2)
        n = n // 2
        if n == 0 : break
    ans = 0
    for i in stack :
        if i == 1 : ans += 1

    return ans

'''
개선된 풀이
def solution(n):
    ans = 0
    while n!=0 :
        if n % 2 == 1 : ans += 1
        n = n // 2
    return ans
'''

File: Programmers/SeungDo/5주차(깊이,너비 우선 탐색(DFS,BFS))/점프와 순간 이동.cpp
using namespace std;
// N부터 0까지 N이 짝수일 때 N/2 홀수일 때 N--,ans++하면서 0까지 순환을 이용하여 풀이해봤습니다.
void cycle(int *n, int* ans) {
    if (*n == 0) return; // 0일때 순환 종료
    if (*n % 2 == 0) *n /= 2; // N이 짝수일 때 순간이동
    else { // N이 홀수일 때 점프후 건전지 사용량 ++
        *n -= 1;
        *ans += 1;
    }
    cycle(n, ans); 
}
int solution(int n)
{
    int ans = 0;
    cycle(&n, &ans); // 순환

    return ans;
}
